codegen/CodeGen/Signature.pm

```diff
@@ -0,0 +1,170 @@
+    if ($_->{multi}) {
+        return "    PTRS_$_->{uc_name} = signature_$_->{name}.search_multiple();\n";
+    }
+    return "    PTR_$_->{uc_name} = signature_$_->{name}.search();\n";
+}
+
+sub yaml_sig_to_cpp_validator {
+    my $validator = "";
+    if ($_->{multi}) {
+        $validator .= "    if (!PTRS_$_->{uc_name}.size())\n";
+    } else {
+        $validator .= "    if (NOT_FOUND(PTR_$_->{uc_name}))\n";
+    }
+
+    if ($_->{crucial}) {
+        $validator .= "        crucial_missing.push_back(signature_$_->{name}.name);\n";
+    } else {
+        $validator .= "        non_crucial_missing.push_back(signature_$_->{name}.name);\n";
+    }
+    return $validator;
+}
+
+sub yaml_sig_to_cpp_address_var {
+    if ($_->{multi} || 0) {
+        return "static std::vector<uintptr_t> PTRS_$_->{uc_name};\n";
+    }
+    return "static uintptr_t PTR_$_->{uc_name};\n";
+}
+
+sub yaml_sig_to_cpp_getter {
+    if ($_->{multi}) {
+        return "std::vector<uintptr_t> sig_$_->{name}() {\n".
+               "    return PTRS_$_->{uc_name};\n".
+               "}\n";
+    }
+    return "$_->{type} sig_$_->{name}() {\n".
+           "    return reinterpret_cast<$_->{type}>(\n".
+           "        NOT_FOUND(PTR_$_->{uc_name}) ?\n".
+           "            0 : (PTR_$_->{uc_name} + $_->{offset}));\n".
+           "}\n";
+
+}
+
+sub yaml_sig_to_cpp_getter_header {
+    if ($_->{multi}) {
+        return "std::vector<uintptr_t> sig_$_->{name}();\n";
+    }
+    return "$_->{type} sig_$_->{name}();\n";
+}
+
+sub yaml_signatures_to_cpp_definitions {
+    my $name = shift;
+    my $sigs = shift;
+    my @sigs = map { preprocess_signature } @{$sigs};
+
+    #### Source file stuff.
+    my $std_includes = [
+        "#include <cstdio>",
+        "#include <cstdint>",
+        "#include <cstdlib>",
+        "#include <vector>",
+        ];
+    my $std_header_includes = [
+        "#include <cstdint>",
+        "#include <vector>",
+        ];
+    my $includes = [
+        "#include <hooker/hooker.hpp>",
+    ];
+    my $source_defs = join "",
+        # Actual signature definitions
+        (map { yaml_sig_to_cpp_sig } @sigs), "\n",
+        # The variables that hold the addresses.
+        (map { yaml_sig_to_cpp_address_var } @sigs), "\n",
+        # Getters for these addresses.
+        (map { yaml_sig_to_cpp_getter } @sigs), "\n";
+
+    my $initialization_code = join "", map { yaml_sig_to_cpp_initializer } @sigs;
+    my $validation_code = join "", map { yaml_sig_to_cpp_validator } @sigs;
+
+    # Function called on initialization.
+    my $init_function = qq{void init_$name\_signatures() {
+    // Write signature results to our variables for safekeeping.
+$initialization_code
+
+    // Find out if we failed to find any signatures.
+    std::vector<const char*> crucial_missing;
+    std::vector<const char*> non_crucial_missing;
+
+$validation_code
+
+    // Report to the user that we are missing signatures if we are.
+    if (crucial_missing.size()) {
```

Comment:
`std::vector` has a function called `empty` that's more idiomatic for what you're doing here.

Reply:
For the empties I'll use empty. but when checking if something is not empty using size seems more appropriate.